crates/svm-runtime-c-api/tests/api_tests.rs

```diff
@@ -1,23 +1,179 @@
+
+    let results = vec![Val::I32(new as i32)];
+    Ok(results)
+}
+
+/// This struct will serve as our `host_env`.
+/// We use here integers as function identifiers to be called by the `trampoline`.
+#[derive(Debug)]
+#[repr(C)]
+struct func_index_t(u32);
+
+type Callback = fn(&mut Context, &[Val]) -> Result<Vec<Val>, RuntimeError>;
+
+const COUNTER_MUL_FN_INDEX: u32 = 123;
+
+fn wasm_trap(err: RuntimeError) -> *mut wasm_trap_t {
+    let trap: wasm_trap_t = err.into();
+
+    // this heap-allocated memory will be released by SVM.
+    // (See: `ExternImport#wasmer_export`)
+
+    Box::into_raw(Box::new(trap))
+}
+
+/// Given a function identifier stored as part of the `host_env`
+/// we can know what Rust native function to call
+fn func_index_to_callback(func_idx: &func_index_t) -> Option<Callback> {
+    match func_idx.0 {
+        COUNTER_MUL_FN_INDEX => Some(counter_mul),
+        _ => None,
+    }
+}
+
+unsafe fn prepare_args(args: *const wasm_val_vec_t) -> Vec<Val> {
+    let args = &*args;
+    let args: &[wasm_val_t] = std::slice::from_raw_parts(args.data, args.size);
+
+    args.iter().map(|v| Val::try_from(v).unwrap()).collect()
+}
+
+/// The memory for results has already by zero-allocated, we're left with filling-in
+/// the values given by `values` parameter.
+unsafe fn copy_results(results: *mut wasm_val_vec_t, values: &[Val]) {
+    let values: Vec<wasm_val_t> = values
+        .iter()
+        .map(|v| wasm_val_t::try_from(v).unwrap())
+        .collect();
+
+    let results: &mut wasm_val_vec_t = &mut *results;
+
+    results.size = values.len();
+
+    for (i, val) in values.iter().enumerate() {
+        let out_ptr = results.data.add(i);
+
+        wasm_val_copy(out_ptr, val);
+    }
+}
+
+/// The `trampoline` is the actual host function that will be called by SVM running.
+/// Each host function will ask SVM to call that `trampoline` function.
+///
+/// The critical difference between different imports is the `host_env` attached to each import.
+/// By placing a function identifier under each `host env` the `trampoline` can figure out to which
+/// Rust function to call. In this example we've introduced `func_index_t` (merely an wrapper for an integer)
+/// as our `host_env`. Other clients can use the same technique very similarly.
+///
+/// Each import function will also give its function signature. The `trampoline` will receive under its values specified
+/// in the import function signature. In addition to required memory for the `results` will be allocated prior calling `trampoline`
+/// such that the `trampoline` will only be left with placing the `results` values.
+///
+/// In case the `trampoline` failed, a pointer to heap-allocated trap will be propagated back to SVM.
+/// SVM will be responsible of deallocating that memory pointed by that `wasm_trap_t`.
+#[no_mangle]
+unsafe extern "C" fn trampoline(
+    env: *mut c_void,
+    args: *const wasm_val_vec_t,
+    results: *mut wasm_val_vec_t,
```

Comment:
The usage of `wasm_val_vec_t` (in `wasm_func_callback_with_env_t` signature) creates a dependency between `svm` client and `wasmer`'s C API. 
This type is being constructed _after_ `wasmer` is calling `svm`'s inner callback, hence `svm` could use a different type and a different signature for `trampoline`.

This type is defined in `wasmer` using a macro. I'm not sure if importing it will require merely using the C headers, or the lib as well. In any case, this would be suboptimal. 

It seems the most straightforward to use `svm`'s `Vec<WasmValue>` (which is also more restrictive in the supported Wasm types). Its encoding format is already defined and used. We can use it, or define `Vec<T>` and `WasmValue` as C types, similarly to how it's done in `wasmer` C API.

Reply:
good point, I've overlooked that fact (I guess I was on an auto-pilot since I'm used to the way things were in the past in the old version of wasmer)